src/main/java/com/example/dog/dao/InMemoryDogDAO.java

```diff
@@ -25,8 +25,9 @@ public Dog add(Dog dog) {
 
     @Override
     public Dog update(int id, Dog dog) {
-        if (!dogs.containsKey(id))
-            return null;
+        if (!dogs.containsKey(id)) {
+            throw new IllegalArgumentException("Dog with id=" + id + " was not found!");
```

Comment:
IllegalArgumentException is too generic. If we want to know that we couldn't find such record, then we should thrown some kind of ObjectNotFoundException. Otherwise you won't be able to distinguish between different IllegalArgumentExceptions when catching them.